openwisp_controller/config/api/generics.py

```diff
@@ -0,0 +1,82 @@
+
+from openwisp_users.models import Organization
+
+
+class BaseCreateTemplateView(CreateAPIView):
+    """
+    API used to create external template.
+    This API will be used at the template library
+    backend repo
+    """
+    authentication_classes = (SessionAuthentication, BasicAuthentication)
+    permission_classes = (DjangoModelPermissions,)
+
+    def save_template(self, template, import_url, request):
+        template.full_clean()
+        template.save()
+        subscriber_url = '{0}://{1}'.format(request.META.get('wsgi.url_scheme'),
+                                            request.get_host())
+        subscribe.delay(template.id, import_url, subscriber_url, subscribe=True)
+
+    def post(self, request, *args, **kwargs):
+        import_url = request.POST.get('import_url', None)
+        data = get_remote_template_data(import_url)
+        org = request.POST.get('org', None)
+        template_org = get_object_or_404(Organization, name=org)
+        data['organization'] = template_org
+        if data['type'] == 'vpn':
+            data['vpn']['ca']['organization'] = template_org
+            data['vpn']['cert']['organization'] = template_org
+            data['vpn']['organization'] = template_org
+        errors = {}
+        if data['type'] == 'vpn':
+            ca = self.ca_model(**data['vpn']['ca'])
+            try:
+                ca.full_clean()
+                ca.save()
+            except ValidationError:
+                ca = self.ca_model.objects.get(pk=ca.id)
+            data['vpn']['cert']['ca'] = ca
+            cert = self.cert_model(**data['vpn']['cert'])
+            try:
+                cert.full_clean()
+                cert.save()
+            except ValidationError:
+                cert = self.cert_model.objects.get(pk=cert.id)
+            data['vpn']['ca'] = ca
+            data['vpn']['cert'] = cert
+            vpn = self.vpn_model(**data['vpn'])
+            try:
+                vpn.full_clean()
+                vpn.save()
+            except ValidationError:
+                vpn = self.vpn_model.objects.get(name=vpn.name)
+            data['vpn'] = vpn
+            template = self.template_model(**data)
+            try:
+                self.save_template(template, import_url, request)
+            except ValidationError as e:
+                errors.update({
+                    'template_errors': e.messages
+                })
+                return Response(data=errors, status=400)
+        else:
+            template = self.template_model(**data)
+            try:
+                self.save_template(template, import_url, request)
+            except ValidationError as e:
+                errors.update({
+                    'template_errors': e.messages
+                })
+                return Response(data=errors, status=400)
+        success = {
+            'template_success': 'Template was successfully created'
+        }
+        return Response(data=success, status=200)
```

Comment:
this looks wrong:

- why is a VPN object being created?
- these operations should be done in a ModelSerializer

Reply:
I had no idea this could be achieve with ModelSerializer. I will look how to achieve this with it. Thanks